core/src/Kokkos_Complex.hpp

```diff
@@ -682,18 +682,58 @@ KOKKOS_INLINE_FUNCTION
       x * y.real(), x * y.imag());
 }
 
+namespace Impl {
+// NOTE promote would also be useful for math functions
+template <class T, bool = std::is_integral<T>::value>
+struct promote {
+  using type = double;
+};
+template <class T>
+struct promote<T, false> {};
+template <>
+struct promote<long double> {
+  using type = long double;
+};
+template <>
+struct promote<double> {
+  using type = double;
+};
+template <>
+struct promote<float> {
+  using type = float;
+};
+template <class T>
+using promote_t = typename promote<T>::type;
+template <class T, class U>
+struct promote_2 {
+  using type = decltype(promote_t<T>() + promote_t<U>());
+};
+template <class T, class U>
+using promote_2_t = typename promote_2<T, U>::type;
+}  // namespace Impl
+
 //! Imaginary part of a complex number.
 template <class RealType>
 KOKKOS_INLINE_FUNCTION RealType imag(const complex<RealType>& x) noexcept {
   return x.imag();
 }
 
+template <class T>
+KOKKOS_INLINE_FUNCTION constexpr Impl::promote_t<T> imag(T) {
+  return T();
+}
```

Comment:
I think it would be helpful to clarify that `T` cannot be arbitrary. Maybe something like
```suggestion
template <class FloatOrInteger>
KOKKOS_INLINE_FUNCTION constexpr Impl::promote_t<FloatOrInteger> imag(FloatInteger) {
  return FloatOrInteger();
}
```
similar to https://en.cppreference.com/w/cpp/numeric/complex/imag2.

Reply:
What about `ArithmeticType`?

Reply:
Sure, that sounds reasonable having https://en.cppreference.com/w/cpp/types/is_arithmetic in mind.